Source/WebCore/platform/graphics/ca/TileController.cpp

```diff
@@ -116,6 +116,114 @@ std::optional<TileGridIdentifier> TileController::secondaryGridIdentifier() cons
+{
+    return m_asyncTileConfigurationChangeData.transform([](const auto& data) {
+        return data.tileConfigurationChange;
+    });
+}
+
+std::optional<TileGridIdentifier> TileController::pendingGridForTileConfigurationChange(const std::optional<TileConfigurationChangeIdentifier>& tileConfigurationChange) const
+{
+    return tileConfigurationChange.and_then([&changeData = m_asyncTileConfigurationChangeData](const auto& change) {
+        return changeData.and_then([&change](const auto& data) -> std::optional<TileGridIdentifier> {
+            if (data.tileConfigurationChange != change)
+                return std::nullopt;
+            return data.pendingTileGrid->identifier();
+        });
+    });
+}
+
+void TileController::destroyExistingTileConfigurationChange()
+{
+    ASSERT(m_asyncTileConfigurationChangeData);
+
+    auto tileConfigurationChange = m_asyncTileConfigurationChangeData->tileConfigurationChange;
+    if (m_client)
+        m_client->didCancelTileConfigurationChange(tileConfigurationChange);
+
+    /*
+     // FIXME: Maybe we want to uncnoditionally DidCancelTileConfigurationChange::No here.
+     auto didCancelTileConfigurationChange = m_asyncTileConfigurationChangeData->pendingTiles.isEmpty() ? DidCancelTileConfigurationChange::No : DidCancelTileConfigurationChange::Yes;
+     m_asyncTileConfigurationChangeData->completionHandler(primaryGridIdentifier(), m_asyncTileConfigurationChangeData->pendingTileGrid->identifier(), didCancelTileConfigurationChange);
+     */
+
+    m_asyncTileConfigurationChangeData->completionHandler(primaryGridIdentifier(), { });
+    m_asyncTileConfigurationChangeData.reset();
+}
+
+void TileController::addPendingTilesToActiveTileConfigurationChangeIfNeeded(const TileGridIndexIteratorRange& tiles)
+{
+    if (!m_asyncTileConfigurationChangeData)
+        return;
+
+    for (const auto& tile : tiles) {
+        auto newTile = *m_asyncTileConfigurationChangeData->pendingTiles.add(tile).iterator;
+        auto tileRect = m_asyncTileConfigurationChangeData->pendingTileGrid->rectForTile(newTile);
+        if (m_client) {
+            m_client->prepareContentForTile(*this, m_asyncTileConfigurationChangeData->pendingTileGrid->identifier(), newTile, tileRect, *activeConfigurationChange(), [newTile, this](auto completedTile) {
+                ASSERT_UNUSED(newTile, newTile == completedTile);
+                didPrepareContentForTile(completedTile);
+            });
+        }
+    }
+}
+
+TileConfigurationChangeIdentifier TileController::initiateTileConfigurationChange(CompletionHandler<void(TileGridIdentifier oldGrid, const std::optional<TileGridIdentifier>& newGrid)>&& completionHandler)
+{
+    if (m_asyncTileConfigurationChangeData)
+        destroyExistingTileConfigurationChange();
+
+    m_asyncTileConfigurationChangeData = { TileConfigurationChangeIdentifier::generate(), makeUnique<TileGrid>(*this), WTFMove(completionHandler), { } };
```

Comment:
Am I right in saying that the GraphicsLayers for the individual tiles will be parented to the controller's GraphicsLayer from both the new and old grid simultaneously?

Is that the behaviour we want? I guess the new tiles can asynchronously change from blank to having content and cover the old tiles, and ordering hopefully means they're on top.